### PR TITLE
401 program event table

### LINKED ISSUE #401: Program status
- [x] table and endpoints
- [x] status in list view
- [x] demo control to schedule events

There are can be various patient status for a program:

- died (generic from patient document)
- ~~opted out; patient stopped program (program data)~~
- Transferred to another clinic (program data)
- active at this clinic (default state no need to store it)
- opted out of treatment  (encounter data)
- interruption of treatment, (encounter data)
- missed appointment (calculated from encounter)
- lost to follow up -> patient hasn't shown up for x days and ran out of medicine

While it is required to show one single program status I think we should not use a single status field to store this information. Reasons are:
- status is gathered from multiple documents (patient, program, and encounters)
- status might not be mutually exclusive, e.g missed appointment and transferred to another clinic.

Alternative is to have a (customizable) view that gathered the available information from different sources and displays it as required. For example, a quite generic status control can be added to the program details view which merges all information into a single status field.

### Lost to follow up
This is the most tricky status since it involves custom logic.
Instead of storing a "lost to follow up" status field we can actually introduce a more generic field from which the "lost to follow up" status can be derived.
A suitable field would be a generic `requiredEncounterDate` field which gives the latest date before the patient must come to the clinic, e.g. to get more critical medicine.
This seems reasonably generic to be included into the encounter base type.
For example, `requiredEncounterDate` can be updated by a custom specialized control that knows the logic how to calculate the next required encounter date before medicine runs out.
`requiredEncounterDate` can be queried in a generic way, e.g. "lost to follow up" programs are programs that have encounters with `requiredEncounterDate < now()`.

To bring this idea even further we can associate a list of events (which are not encounter events) with an encounter. For example (2020_08_08 now):
- 2020_11_11: vaccine becomes active/ patient is now immune
- 2020_12_01: next dentist appointment due
- 2022_01_15: lost to follow up

### PR DESCRIPTION
**Pull request by clemens-msupply**
- Adds program event table (this table is reasonable generic enough to be used for other purposes in the future)
- Service + API to update events and to query them
- DispensedDuration Control to schedule events based on the control's "number of days dispensed" value

Testing:
- reinit the DB
- Program for Tina will show status "Lost to follow up"
- Go to one of her encounters and set the "Number of  Days Dispensed" field to a value > 0 (there are two controls for it use the top one)
- Save the encounter
- Status should now be cleared
- wait 1 minute -> status change to "Interupted", wait another 2 minutes -> status changes to "Lost to follow up"

Some program event table motivation:
- Events have a patient + context which can be used to assign them, for example, to a program enrolment or some other entity 
- Within this context events can be grouped (group field). This is useful to replace a group of events with an updated list of events in this group, leaving other events in the context untouched.

Closes #401

### CODE REVIEW

File: server/service/src/sync/init_programs_data.rs [hunk 1]
```diff
@@ -323,6 +322,26 @@ fn encounter_2() -> hiv_care_encounter::HivcareEncounter {
...
+                name: Some("Interrupted".to_string()),
+            },
+            hiv_care_encounter::EncounterEvent {
+                datetime: time
+                    .checked_add_signed(Duration::weeks(2))
+                    .unwrap()
+                    .to_rfc3339(),
+                group: Some("HivCareEncounterDispensingStatus".to_string()),
+                type_: "status".to_string(),
+                name: Some("Lost".to_string()),
```
Comment (mark-prins) [inline server/service/src/sync/init_programs_data.rs:342]: should this be `Lost to follow up` to match the 'number of days dispensed' event?

**Review by mark-prins — APPROVED** [review]
🎉  that's awesome! a lot of thought has gone into that one PR. I probably haven't done the server aspects justice in the review: but have tested via f/e and tests running against postgres. All is working nicely.

File: client/packages/system/src/Patient/ProgramEnrolment/Components/ListView.tsx [hunk 2]
```diff
@@ -37,6 +37,11 @@ const ProgramListComponent: FC = () => {
         key: 'programPatientId',
         label: 'label.enrolment-patient-id',
       },
+      {
+        key: 'events',
+        label: 'label.status',
+        formatter: events => (events as { name?: string }[])[0]?.name ?? '',
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/system/src/Patient/ProgramEnrolment/Components/ListView.tsx:43]: Yes got a bit sloppy, I refactored it a bit

File: client/packages/system/src/Patient/JsonForms/components/DispendedDuration.tsx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+const MINUTE = 60 * 1000;
+const DAY = 24 * 60 * MINUTE;
+
+const scheduleEvent = (
+  event: OptionEvent,
+  baseTime: number
+): EncounterEvent => {
+  const datetime =
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/system/src/Patient/JsonForms/components/DispendedDuration.tsx:70]: I am using addDays and addMinutes now, which also removed the need for the time constants

File: client/packages/system/src/Patient/JsonForms/components/DispendedDuration.tsx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      let events: EncounterEvent[] = [];
+      if (value > 0) {
+        const baseTimeField = extractProperty(
+          data,
+          options?.baseDatetimeField ?? ''
+        );
+
+        const scheduleStartTime = options.scheduleEventsNow
+          ? Date.now()
+          : new Date(baseTimeField).getTime() + value * DAY;
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/system/src/Patient/JsonForms/components/DispendedDuration.tsx:113]: Done, I also thought its better to use the startOfDay. Doing this I also made the baseTime extraction more robust (when creating a new encounter the startDatetime could currently be undefined).

**Comment by clemens-msupply** [comment]
Thanks! I implemented the suggested changes and fixed some bugs while doing this...

**Pull request merged**